dolphin/db/sqlalchemy/api.py

```diff
@@ -305,9 +305,98 @@ def storage_delete(context, storage_id):
     _storage_get_query(context).filter_by(id=storage_id).delete()
 
 
+def _volume_get_query(context, session=None):
+    return model_query(context, models.Volume, session=session)
+
+
+def _volume_get(context, volume_id, session=None):
+    result = (_volume_get_query(context, session=session)
+              .filter_by(id=volume_id)
+              .first())
+
+    if not result:
+        raise exception.VolumeNotFound(id=volume_id)
+
+    return result
+
+
+def volume_create(context, values):
+    """Create a volume."""
+    if not values.get('id'):
+        values['id'] = uuidutils.generate_uuid()
+
+    vol_ref = models.Volume()
+    vol_ref.update(values)
+
+    session = get_session()
+    with session.begin():
+        session.add(vol_ref)
+
+    return _volume_get(context,
+                       vol_ref['id'],
+                       session=session)
+
+
+def volumes_create(context, volumes):
+    """Create multiple volumes."""
+    session = get_session()
+    vol_refs = []
+    with session.begin():
+
+        for vol in volumes:
+            LOG.debug('adding new volume for original_id {0}:'
+                      .format(vol.get('original_id')))
+            if not vol.get('id'):
+                vol['id'] = uuidutils.generate_uuid()
+
+            vol_ref = models.Volume()
+            vol_ref.update(vol)
+            vol_refs.append(vol_ref)
+
+        session.add_all(vol_refs)
+
+    return vol_refs
+
+
+def volumes_delete(context, volumes_id_list):
+    """Delete multiple volumes."""
+    session = get_session()
+    with session.begin():
+        for vol_id in volumes_id_list:
+            LOG.debug('deleting volume {0}:'.format(vol_id))
+            query = _volume_get_query(context, session)
+            result = query.filter_by(id=vol_id).delete()
+
+            if not result:
+                LOG.error(exception.VolumeNotFound(id=vol_id))
+    return
+
+
+def volume_update(context, vol_id, values):
+    """Update a volume."""
+    session = get_session()
+    with session.begin():
+        _volume_get(context, vol_id, session).update(values)
+    return _volume_get(context, vol_id, session)
+
+
+def volumes_update(context, volumes):
+    """Update multiple volumes."""
+    session = get_session()
+    with session.begin():
+        for vol in volumes:
+            LOG.debug('updating volume {0}:'.format(vol.get('id')))
+            query = _volume_get_query(context, session)
+            result = query.filter_by(id=vol.get('id')
+                                     ).update(vol)
+
+            if not result:
+                LOG.debug(exception.VolumeNotFound(id=vol.get('id')))
```

Comment:
Suggest to use error, cause it's actually a kind of exception.

Reply:
done